key.go

```diff
@@ -0,0 +1,810 @@
+package cose
+
+import (
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/ed25519"
+	"crypto/elliptic"
+	"errors"
+	"fmt"
+	"math/big"
+	"strconv"
+
+	cbor "github.com/fxamacker/cbor/v2"
+)
+
+const (
+	// An inviald key_op value
+	KeyOpInvalid KeyOp = 0
+
+	// The key is used to create signatures. Requires private key fields.
+	KeyOpSign KeyOp = 1
+
+	// The key is used for verification of signatures.
+	KeyOpVerify KeyOp = 2
+
+	// The key is used for key transport encryption.
+	KeyOpEncrypt KeyOp = 3
+
+	// The key is used for key transport decryption. Requires private key fields.
+	KeyOpDecrypt KeyOp = 4
+
+	// The key is used for key wrap encryption.
+	KeyOpWrapKey KeyOp = 5
+
+	// The key is used for key wrap decryption.
+	KeyOpUnwrapKey KeyOp = 6
+
+	// The key is used for deriving keys. Requires private key fields.
+	KeyOpDeriveKey KeyOp = 7
+
+	// The key is used for deriving bits not to be used as a key. Requires
+	// private key fields.
+	KeyOpDeriveBits KeyOp = 8
+
+	// The key is used for creating MACs.
+	KeyOpMACCreate KeyOp = 9
+
+	// The key is used for validating MACs.
+	KeyOpMACVerify KeyOp = 10
+)
+
+// KeyOp represents a key_ops value used to restrict purposes for which a Key
+// may be used.
+type KeyOp int64
+
+// KeyOpFromString returns the KeyOp corresponding to the specified name.
+// The values are taken from https://www.rfc-editor.org/rfc/rfc7517#section-4.3
+func KeyOpFromString(val string) (KeyOp, error) {
+	switch val {
+	case "sign":
+		return KeyOpSign, nil
+	case "verify":
+		return KeyOpVerify, nil
+	case "encrypt":
+		return KeyOpEncrypt, nil
+	case "decrypt":
+		return KeyOpDecrypt, nil
+	case "wrapKey":
+		return KeyOpWrapKey, nil
+	case "unwrapKey":
+		return KeyOpUnwrapKey, nil
+	case "deriveKey":
+		return KeyOpDeriveKey, nil
+	case "deriveBits":
+		return KeyOpDeriveBits, nil
+	default:
+		return KeyOpInvalid, fmt.Errorf("unknown key_ops value %q", val)
+	}
+}
+
+func (ko KeyOp) String() string {
+	switch ko {
+	case KeyOpSign:
+		return "sign"
+	case KeyOpVerify:
+		return "verify"
+	case KeyOpEncrypt:
+		return "encrypt"
+	case KeyOpDecrypt:
+		return "decrypt"
+	case KeyOpWrapKey:
+		return "wrapKey"
+	case KeyOpUnwrapKey:
+		return "unwrapKey"
+	case KeyOpDeriveKey:
+		return "deriveKey"
+	case KeyOpDeriveBits:
+		return "deriveBits"
+	case KeyOpMACCreate:
+		return "MAC create"
+	case KeyOpMACVerify:
+		return "MAC verify"
```

Comment:
The strings for MAC create and verify seem not defined.

Reply:
Given that no string value is defined these, I'm using the `Name` here instead. `String()` is not used during serialization to CBOR, and does not represent the encoded value, it only provides a string reprsentation (for use in error messages, debug output, etc).

Reply:
To prevent misuse, can you document the special cases of "MAC create/verify"?